src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs

```diff
@@ -89,5 +90,16 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
                 return snapshot.GetOpenDocumentInCurrentContextWithChanges();
             }
         }
+
+        /// <summary>
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetOpenDocumentInCurrentContextWithChanges(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns
+        /// </summary>
+        public static Document GetFullyLoadedOpenDocumentInCurrentContextWithChanges(
```

Comment:
❗️ This method had the correct signature _before_ this change. Sync/async transitions should always be pushed to the callers, not to the callees.

Reply:
"The transition should always be moved as far to the callers as possible" is true in normal Task case since Wait on a task could lead to deadlock, VS or not. especially due to UI thread in VS since UI thread act as a single lock.

I am not sure whether that is true for JTF since one of the purposes of JTF is to hide such detail I believe. in practice, that might be the whole purpose of people why they are using it.

otherwise, none of VS APIs, I mean literally none of APIs we use that internally use JTF should not use it. since that basically violates what you said above. we currently have no idea which VS API we uses internally use JTF (especially JTF.Run) and asked just use JTF always to make whole thing works in VS.

if we are in the world you stated above, we should know exactly what VS APIs are using JTF inside and we are required to use them under JTF in UI thread.

but we don't live in such world. and Roslyn which doesn't adapt JTF fully make things more complicated. 

we already have many APIs which support sync and async version of APIs such as Read and ReadAsync or GetRoot and GetRootAsync and more.

I want people to consider this API as such. sure, people might use sync version in async method and then call Wait on the async method causing us a deadlock. but conceptually same problem already exist in other cases such as sync method calling async version of GetRootAsync which will cause thread starvation. and we provide sync API so that caller has an option not to call async version in sync code path.

I would like users to use it that way. they don't need to know about the implementation detail and as long as they use sync version in sync code path, they are fine. sure there is always corner case where sync code path is nested deep in sync code path that got called with async method with async method at the top with Wait.

to help avoiding such case, it is an extension method and actual service doesn't have sync API. also, I dont think the concern above can be removed by forcing people to repeatedly write the same JTF.Run code. now they have the risk of not calling this in JTF.Run and cause hang.

on whether we should always use JTF.Run and never wait. that is not how Roslyn is written. and until we get rid of all Wait from Roslyn, Wait/Result is valid option for Roslyn.